enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,756 @@
+This restriction is known and will be documented.
+Additional features provided by Cluster API will not be ported to Machine API as part of this effort.
+
+##### ControlPlaneMachineSet Synchronisation
+
+ControlPlaneMachineSets are not present in Cluster API.
+However, the ControlPlaneMachineSet was designed in such a way that we can extend the `template`` section (which is a discriminated union) to support Cluster API.
+
+A separate enhancement will be created for the support of Cluster API in the ControlPlaneMachineSet, however, for the sake of conversion,
+we expect the conversion to happen much in the same way as a regular MachineSet.
+
+ControlPlaneMachineSets (additionally to MachineSets) have a concept of failure domains.
+This is also a concept in Cluster API.
+When converting ControlPlaneMachineSets to Cluster API, we will ensure that the appropriate failure domain configuration is configured inside the Cluster API InfrastructureCluster,
+and that the failure domains from the ControlPlaneMachineSet Machine API configuration are reflected across.
+
+When deploying Machines using the future Cluster API extension, the InfrastructureTemplate will be combined with the named failure domains on the InfrastructureCluster
+to spread the Machines across multiple failure domains.
+
+#### Creation of new Machines
+
+When a Machine API MachineSet creates a new Machine, the synchronisation controller will create the equivalent Cluster API Machine based on the method described in [Machine Synchronisation](#machine-synchronisation).
+
+When a Cluster API MachineSet creates a new Machine, the synchronisation controller will create it’s Machine API equivalent if, and only if, the Cluster API MachineSet has a Machine API equivalent.
+This will allow users to switch a MachineSet back to Machine API at a later date if they wish to do so.
+
+When the Cluster API MachineSet has been created independently of Machine API, new Machines will not be created in Machine API to reflect the Machines owned by the Cluster API MachineSet.
+
+#### Deletion of Machines
+
+To enable the synchronisation of deletion, the synchronisation controller will use its own Finalizer on resources that have Machine API and Cluster API equivalents.
+
+This means that for a Machine, we expect the authoritative controller to perform the actual deletion work within the infrastructure provider.
+Once it has completed this effort, it will remove its own Finalizer.
+
+As soon as the synchronisation controller notices the Machine has been deleted, it should ensure that the Machine API/Cluster API equivalent (if present) is also deleted.
+It should then wait until the authoritative Machine controller removes its finalizer before removing both its own, and the Machine API/Cluster API Machine finalizer if present.
```

Comment:
Yes I think the intention here was that it would remove the non-authoritative controller finalizer if it were added. The reason for that being that we want to, as minimally as possible, add code to the controllers to handle being non-authoritative. If we prevent them reconciling at all, they _should_ not add their finalizer unless they were at some point authoritative. But, when non-authoritative, they shouldn't execute their deletion routine as they shouldn't be reconciling, meaning something else has to remove it.

I agree in general this is bad practice, but since we have two controllers owning the same cloud resource, the finalizers represent the same action, so in theory this should be safe as described

Reply:
I mean... the non-authoritative reconcile has to run to _some_ degree, if only to decide it shouldn't reconcile. It gives me less heartburn for it to have logic like

```
if me.authoritative:
  ensureFinalizerPresent()
  // and carry on
else:
  ensureFinalizerAbsent()
  return nil
```

than for a two different controllers to be managing the same finalizer.

Reply:
> the non-authoritative reconcile has to run to some degree

Possibly, and if it ends up like that then you're right, we have this option. I have an open question to explore how the filtering will be done, but my hope was to filter events using predicates before we even get to `Reconcile`, but again, needs testing.